c/examples/connection_retry.c

```diff
@@ -0,0 +1,123 @@
+/**************************************************************************/ /**
+ * @brief Example Polaris client with reconnect/reauthentication support.
+ *
+ * Copyright (c) Point One Navigation - All Rights Reserved
+ ******************************************************************************/
+
+#include <signal.h>
+
+#include "point_one/polaris/polaris.h"
+#include "point_one/polaris/portability.h"
+
+static PolarisContext_t context;
+
+void HandleData(const uint8_t* buffer, size_t size_bytes) {
+  P1_printf("Application received %u bytes.\n", (unsigned)size_bytes);
+}
+
+void HandleSignal(int sig) {
+  signal(sig, SIG_DFL);
+
+  P1_printf("Caught signal %s (%d). Closing Polaris connection.\n",
+            sys_siglist[sig], sig);
+  Polaris_Disconnect(&context);
+}
+
+int main(int argc, const char* argv[]) {
+  if (argc != 3) {
+    P1_fprintf(stderr, "Usage: %s API_KEY UNIQUE_ID\n", argv[0]);
+    return 1;
+  }
+
+  const char* api_key = argv[1];
+  const char* unique_id = argv[2];
+
+  const int MAX_RECONNECTS = 2;
+  int auth_valid = 0;
+  int reconnect_count = 0;
+  while (1) {
+    // Retrieve an access token using the specified API key.
+    if (!auth_valid) {
+      if (Polaris_Init(&context) != POLARIS_SUCCESS) {
+        return 2;
+      }
+
+      P1_printf("Opened Polaris context. Authenticating...\n");
+
+      int ret = Polaris_Authenticate(&context, api_key, unique_id);
+      if (ret == POLARIS_FORBIDDEN) {
+        P1_printf("Authentication rejected. Is your API key valid?\n");
+        return 3;
+      } else if (ret != POLARIS_SUCCESS) {
+        continue;
+      }
+
+      auth_valid = 1;
+    }
+
+    // We now have a valid access token. Connect to the corrections service.
+    //
+    // In the calls below, if the connection times out or the access token is
+    // rejected, try to connect again. If it fails too many times, the access
+    // token may be expired - try reauthenticating.
+    P1_printf("Authenticated. Connecting to Polaris...\n");
+
+    Polaris_SetRTCMCallback(&context, HandleData);
+
+    if (Polaris_Connect(&context) != POLARIS_SUCCESS) {
+      if (++reconnect_count >= MAX_RECONNECTS) {
+        P1_printf(
+            "Max reconnects exceeded. Clearing access token and retrying "
+            "authentication.\n");
+        auth_valid = 0;
+        reconnect_count = 0;
+      }
+      continue;
+    }
+
+    P1_printf("Connected to Polaris...\n");
+
+    // Send a position update to Polaris. Position updates are used to select an
+    // appropriate corrections stream, and should be updated periodically as the
+    // receiver moves.
+    if (Polaris_SendECEFPosition(&context, -2707071.0, -4260565.0, 3885644.0) !=
```

Comment:
Meant to say these should be cmd args

Reply:
Yeah I didn't want to have to implement getopt or some other complicated argument parsing since this is just an example application that people can modify if they want. They already have to pass in 2 arguments (API key and unique ID), so passing in 3 more, or a comma-separated value or something, seemed unnecessarily annoying.